SpatialGDK/Source/SpatialGDK/Public/Interop/RPCs/RPCTypes.h

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "Interop/Connection/SpatialGDKSpanId.h"
+#include "Schema/RPCPayload.h"
+#include "SpatialView/EntityView.h"
+
+namespace SpatialGDK
+{
+struct RPCReadingContext
+{
+	Worker_EntityId EntityId;
+	Worker_ComponentId ComponentId;
+
+	bool IsUpdate() const { return Update != nullptr; }
+
+	Schema_ComponentUpdate* Update = nullptr;
+	Schema_Object* Fields = nullptr;
+};
+
+struct RPCWritingContext
+{
+	enum class DataKind
+	{
+		Generic,
+		ComponentData,
+		ComponentUpdate,
+		CommandRequest,
+		CommandResponse
+	};
+
+	class EntityWrite
+	{
+	public:
+		EntityWrite(const EntityWrite&) = delete;
+		EntityWrite& operator=(const EntityWrite&) = delete;
+		EntityWrite& operator=(EntityWrite&&) = delete;
+
+		EntityWrite(EntityWrite&& Write);
+		~EntityWrite();
+
+		Schema_ComponentUpdate* GetComponentUpdateToWrite();
+		Schema_Object* GetFieldsToWrite();
+
+		void RPCWritten(uint32 RPCId);
+
+		const Worker_EntityId EntityId;
+		const Worker_ComponentId ComponentId;
+
+	private:
+		union
+		{
+			Schema_GenericData* GenData;
+			Schema_ComponentData* Data;
+			Schema_ComponentUpdate* Update;
+			Schema_CommandRequest* Request;
+			Schema_CommandResponse* Response;
+		};
+
+		EntityWrite(RPCWritingContext& InCtx, Worker_EntityId InEntityId, Worker_ComponentId InComponentID);
+		friend RPCWritingContext;
```

Comment:
Nit - I don't think you need this. If it's nested the parent can access the private fields.

Reply:
IIRC it's the other way around.